File: test/go/disbursement/finish_notify_test.go
package disbursement_test

import (
	"context"
	"encoding/json"
	"fmt"
	"testing"
	"time"

	"github.com/dana-id/dana-go/v2/exceptions"
	pg "github.com/dana-id/dana-go/v2/payment_gateway/v1"
	"github.com/google/uuid"

	"uat-script/helper"
)

const (
	createOrderTitleCaseFinishNotify             = "CreateOrder"
	createOrderJsonPathFinishNotify              = "../../../resource/request/components/PaymentGateway.json"
	createOrderRequestCaseFinishNotify           = "CreateOrderApi"
	createOrderAssertCaseFinishNotify            = "CreateOrderApi"
	notificationN8nURL                           = "https://n8n.automation.dana.id/webhook/3676a08f-b06e-416c-b6cd-bea04f71c4d5"
	finishNotifyDefaultValidUpToOffsetSeconds    = 360
	finishNotifyValidUpToOffsetExpiredSeconds    = 2*60 + 15
)

func patchCreateOrderAPIForFinishNotify(jsonDict map[string]interface{}, amount string) {
	if amt, ok := jsonDict["amount"].(map[string]interface{}); ok {
		amt["value"] = amount
	}
	jsonDict["payOptionDetails"] = []interface{}{
		map[string]interface{}{
			"payMethod": "VIRTUAL_ACCOUNT",
			"payOption": "VIRTUAL_ACCOUNT_CIMB",
			"transAmount": map[string]interface{}{
				"value":    amount,
				"currency": "IDR",
			},
		},
	}
	if ups, ok := jsonDict["urlParams"].([]interface{}); ok {
		for _, u := range ups {
			if um, ok := u.(map[string]interface{}); ok {
				if typ, ok := um["type"].(string); ok && typ == "NOTIFICATION" {
					um["url"] = notificationN8nURL
				}
			}
		}
	}
}

func formatCreateOrderAPIError(err error) string {
	if err == nil {
		return ""
	}
	if apiErr, ok := err.(*exceptions.GenericOpenAPIError); ok {
		if body := string(apiErr.Body()); body != "" {
			return fmt.Sprintf("%v: %s", err, body)
		}
	}
	return err.Error()
}

func payVAFromCreateOrderResponse(t *testing.T, body string) {
	t.Helper()
	paymentCode, err := helper.PaymentCodeFromCreateOrderResponse(body)
	if err != nil {
		t.Fatalf("extract paymentCode: %v", err)
	}
	if err := helper.PayVirtualAccountSandbox(paymentCode); err != nil {
		t.Fatalf("pay VA: %v", err)
	}
}

func createOrderAPIFinishNotifyOnce(amount, validUpTo string) (partnerRef string, responseJSON string, err error) {
	result, err := helper.RetryOnInconsistentRequest(func() (interface{}, error) {
		jsonDict, err := helper.GetRequest(createOrderJsonPathFinishNotify, createOrderTitleCaseFinishNotify, createOrderRequestCaseFinishNotify)
		if err != nil {
			return nil, err
		}

		partnerRef := uuid.New().String()
		jsonDict["partnerReferenceNo"] = partnerRef
		if validUpTo != "" {
			jsonDict["validUpTo"] = validUpTo
		} else {
			jsonDict["validUpTo"] = helper.GenerateFormattedDate(finishNotifyDefaultValidUpToOffsetSeconds, 7)
		}
		patchCreateOrderAPIForFinishNotify(jsonDict, amount)

		createOrderByApiRequest := &pg.CreateOrderByApiRequest{}
		jsonBytes, err := json.Marshal(jsonDict)
		if err != nil {
			return nil, err
		}
		if err := json.Unmarshal(jsonBytes, createOrderByApiRequest); err != nil {
			return nil, err
		}

		createOrderReq := pg.CreateOrderRequest{
			CreateOrderByApiRequest: createOrderByApiRequest,
		}

		ctx := context.Background()
		apiResponse, httpResponse, err := helper.ApiClient.PaymentGatewayAPI.CreateOrder(ctx).CreateOrderRequest(createOrderReq).Execute()
		if err != nil {
			return nil, err
		}
		defer httpResponse.Body.Close()

		out, err := apiResponse.MarshalJSON()
		if err != nil {
			return nil, err
		}
		return map[string]interface{}{
			"partnerRef": partnerRef,
			"body":       string(out),
		}, nil
	}, 3, 2*time.Second)
	if err != nil {
		return "", "", err
	}
	if result == nil {
		return "", "", nil
	}
	m := result.(map[string]interface{})
	return m["partnerRef"].(string), m["body"].(string), nil
}

func TestTransactionSuccessNotify(t *testing.T) {
	partnerRef, body, err := createOrderAPIFinishNotifyOnce("11011.00", "")
	if err != nil {
		t.Fatalf("create order: %s", formatCreateOrderAPIError(err))
	}
	if partnerRef == "" || body == "" {
		t.Fatal("empty partner reference or response")
	}
	if err := helper.AssertResponse(createOrderJsonPathFinishNotify, createOrderTitleCaseFinishNotify, createOrderAssertCaseFinishNotify, body, map[string]interface{}{
		"partnerReferenceNo": partnerRef,
	}); err != nil {
		t.Fatal(err)
	}
	payVAFromCreateOrderResponse(t, body)
}

func TestInternalServerErrorNotify(t *testing.T) {
	partnerRef, body, err := createOrderAPIFinishNotifyOnce("11012.00", "")
	if err != nil {
		t.Fatalf("create order: %s", formatCreateOrderAPIError(err))
	}
	if partnerRef == "" || body == "" {
		t.Fatal("empty partner reference or response")
	}
	if err := helper.AssertResponse(createOrderJsonPathFinishNotify, createOrderTitleCaseFinishNotify, createOrderAssertCaseFinishNotify, body, map[string]interface{}{
		"partnerReferenceNo": partnerRef,
	}); err != nil {
		t.Fatal(err)
	}
	payVAFromCreateOrderResponse(t, body)
}

func TestExpiredNotify(t *testing.T) {
	validUpTo := helper.GenerateFormattedDate(finishNotifyValidUpToOffsetExpiredSeconds, 7)
	partnerRef, body, err := createOrderAPIFinishNotifyOnce("11013.00", validUpTo)
	if err != nil {
		t.Fatalf("create order: %s", formatCreateOrderAPIError(err))
	}
	if partnerRef == "" || body == "" {
		t.Fatal("empty partner reference or response")
	}
	if err := helper.AssertResponse(createOrderJsonPathFinishNotify, createOrderTitleCaseFinishNotify, createOrderAssertCaseFinishNotify, body, map[string]interface{}{
		"partnerReferenceNo": partnerRef,
	}); err != nil {
		t.Fatal(err)
	}
}

File: test/java/src/test/java/id/dana/disbursement/FinishNotifyTest.java
package id.dana.disbursement;

import static org.junit.jupiter.api.Assertions.fail;

import id.dana.invoker.Dana;
import id.dana.invoker.model.DanaConfig;
import id.dana.invoker.model.constant.EnvKey;
import id.dana.invoker.model.enumeration.DanaEnvironment;
import id.dana.paymentgateway.PaymentPGUtil;
import id.dana.paymentgateway.v1.api.PaymentGatewayApi;
import id.dana.paymentgateway.v1.model.CreateOrderByApiRequest;
import id.dana.paymentgateway.v1.model.CreateOrderResponse;
import id.dana.paymentgateway.v1.model.Money;
import id.dana.paymentgateway.v1.model.PayOptionDetail;
import id.dana.paymentgateway.v1.model.PayOptionDetail.PayMethodEnum;
import id.dana.paymentgateway.v1.model.PayOptionDetail.PayOptionEnum;
import id.dana.paymentgateway.v1.model.UrlParam;
import id.dana.util.ConfigUtil;
import id.dana.util.RetryTestUtil.Retry;
import id.dana.util.TestUtil;

import java.util.Collections;
import java.util.HashMap;
import java.util.Map;
import java.util.UUID;

import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class FinishNotifyTest extends AbstractDisbursementTest {

  private static final Logger log = LoggerFactory.getLogger(FinishNotifyTest.class);

  private static final String titleCase = "CreateOrder";
  private static final String createOrderRequestCaseFinishNotify = "CreateOrderApi";
  private static final String createOrderAssertCaseFinishNotify = "CreateOrderApi";
  private static final String notificationN8nURL =
      "https://n8n.automation.dana.id/webhook/3676a08f-b06e-416c-b6cd-bea04f71c4d5";
  private static final long finishNotifyDefaultValidUpToOffsetSeconds = 360;
  private static final long finishNotifyValidUpToOffsetExpiredSeconds = 2 * 60 + 15;
  private static final String jsonPathFile = FinishNotifyTest.class.getResource(
      "/request/components/PaymentGateway.json").getPath();
  private static final String merchantId = ConfigUtil.getConfig("MERCHANT_ID", "216620010016033632482");

  private PaymentGatewayApi api;

  @BeforeEach
  void setUp() {
    DanaConfig.Builder danaConfigBuilder = new DanaConfig.Builder();
    danaConfigBuilder
        .partnerId(ConfigUtil.getConfig("X_PARTNER_ID", ""))
        .privateKey(ConfigUtil.getConfig("PRIVATE_KEY", ""))
        .origin(ConfigUtil.getConfig("ORIGIN", ""))
        .env(DanaEnvironment.getByName(ConfigUtil.getConfig(EnvKey.ENV, "SANDBOX")));

    DanaConfig.getInstance(danaConfigBuilder);
    api = Dana.getInstance().getPaymentGatewayApi();
  }

  private void patchCreateOrderAPIForFinishNotify(CreateOrderByApiRequest requestData, String amount) {
    if (requestData.getAmount() != null) {
      requestData.getAmount().setValue(amount);
    }
    Money transAmount = new Money();
    transAmount.setValue(amount);
    transAmount.setCurrency("IDR");
    PayOptionDetail vaDetail = new PayOptionDetail();
    vaDetail.setPayMethod(PayMethodEnum.VIRTUAL_ACCOUNT);
    vaDetail.setPayOption(PayOptionEnum.VIRTUAL_ACCOUNT_CIMB);
    vaDetail.setTransAmount(transAmount);
    requestData.setPayOptionDetails(Collections.singletonList(vaDetail));
    if (requestData.getUrlParams() != null) {
      for (UrlParam u : requestData.getUrlParams()) {
        if (u != null && "NOTIFICATION".equals(u.getType())) {
          u.setUrl(notificationN8nURL);
        }
      }
    }
  }

  private Map<String, Object> createOrderAPIFinishNotifyOnce(String amount, String validUpTo) {
    CreateOrderByApiRequest requestData = PaymentPGUtil.getCreateOrderApiRequest(
        jsonPathFile, titleCase, createOrderRequestCaseFinishNotify);

    String partnerReferenceNo = UUID.randomUUID().toString();
    requestData.setPartnerReferenceNo(partnerReferenceNo);
    requestData.setMerchantId(merchantId);
    if (validUpTo != null && !validUpTo.isEmpty()) {
      requestData.setValidUpTo(validUpTo);
    } else {
      requestData.setValidUpTo(PaymentPGUtil.generateDateWithOffsetSeconds(
          finishNotifyDefaultValidUpToOffsetSeconds));
    }
    patchCreateOrderAPIForFinishNotify(requestData, amount);

    try {
      CreateOrderResponse response = api.createOrder(requestData);
      Map<String, Object> result = new HashMap<>();
      result.put("partnerReferenceNo", partnerReferenceNo);
      result.put("response", response);
      return result;
    } catch (Exception e) {
      throw new RuntimeException(e);
    }
  }

  @Test
  @Retry(value = 3, waitMs = 2000)
  void testTransactionSuccessNotify() {
    runFinishNotifyTest("11011.00", "", true);
  }

  @Test
  @Retry(value = 3, waitMs = 2000)
  void testInternalServerErrorNotify() {
    runFinishNotifyTest("11012.00", "", true);
  }

  @Test
  @Retry(value = 3, waitMs = 2000)
  void testExpiredNotify() {
    String validUpTo = PaymentPGUtil.generateDateWithOffsetSeconds(
        finishNotifyValidUpToOffsetExpiredSeconds);
    runFinishNotifyTest("11013.00", validUpTo, false);
  }

  private void runFinishNotifyTest(String amount, String validUpTo, boolean payVA) {
    try {
      Map<String, Object> result = createOrderAPIFinishNotifyOnce(amount, validUpTo);
      String partnerReferenceNo = (String) result.get("partnerReferenceNo");
      CreateOrderResponse response = (CreateOrderResponse) result.get("response");

      Map<String, Object> variableDict = new HashMap<>();
      variableDict.put("partnerReferenceNo", partnerReferenceNo);

      TestUtil.assertResponse(jsonPathFile, titleCase, createOrderAssertCaseFinishNotify, response, variableDict);

      if (payVA) {
        PaymentPGUtil.payVirtualAccountSandbox(
            PaymentPGUtil.paymentCodeFromCreateOrderResponse(response));
      }
    } catch (Exception e) {
      log.error("Finish notify create order test failed:", e);
      fail("Finish notify create order test failed: " + e.getMessage());
    }
  }
}
